core/src/test/java/io/aiven/kafka/tieredstorage/chunkmanager/cache/ChunkCacheTest.java

```diff
@@ -336,5 +336,47 @@ void failedReadingCachedValueWithExecutionException() throws Exception {
                     .hasRootCauseInstanceOf(RuntimeException.class)
                     .hasRootCauseMessage(TEST_EXCEPTION_MESSAGE);
         }
+
+        @Test
+        void cacheIsNotPoisonedWithFailedFuturesOnFetching() throws Exception {
+            when(chunkManager.getChunk(eq(SEGMENT_OBJECT_KEY), eq(SEGMENT_MANIFEST), eq(0)))
+                .thenThrow(new StorageBackendException(TEST_EXCEPTION_MESSAGE))
+                .thenReturn(new ByteArrayInputStream(new byte[1]));
+
+            assertThatThrownBy(() -> chunkCache
+                .getChunk(SEGMENT_OBJECT_KEY, SEGMENT_MANIFEST, 0))
+                .isInstanceOf(StorageBackendException.class)
+                .hasMessage(TEST_EXCEPTION_MESSAGE);
+
+            await().atMost(Duration.ofMillis(50))
+                .pollInterval(Duration.ofMillis(5))
+                .ignoreExceptions()
+                .until(() ->
+                    chunkCache.getChunk(SEGMENT_OBJECT_KEY, SEGMENT_MANIFEST, 0).readAllBytes().length == 1);
```

Comment:
Do we really need it like this? I would expect an exception to be never thrown here. more over it's a second call to mock that should always return 1 byte as specified in thenReturn. So should it be just
`assertThat(chunkCache.getChunk(SEGMENT_OBJECT_KEY, SEGMENT_MANIFEST, 0)).hasBinaryContent(new byte[1]);` ?

Reply:
Also instead of plain comparison in `until` it would be cleaner to user `untilAsserted` and use AssertJ assertion inside.

Reply:
The removal of the failed future is asynchronous and it becomes flaky without waiting.

> Also instead of plain comparison in `until` it would be cleaner to user `untilAsserted` and use AssertJ assertion inside.

Good point, I'll make a PR.